### PR TITLE
Clear witness cache when re-witnessing notes

### LINKED ISSUE #1378: error on startup: Assertion `nWitnessCacheSize >= nd->witnesses.size()' failed.
```
UpdateTip: new best=00311e4f2dae18b682086110e4ae2581857a0bb31e9eb66c127223aebffb9638  height=621  log2_work=17.079724  tx=632  date=2016-09-10 11:44:35 progress=1.000000  cache=0.2MiB(535tx)
AddToWallet 09d48c0a92b3a993ccc538555afe23388edc4c4743e473cd290a5bc6e0f6cafc  
UpdateTip: new best=0064a254e4bc19fbae2d8ef31761cfe3c2c529cd18b24d6578979e395069ba26  height=622  log2_work=17.085939  tx=634  date=2016-09-10 11:47:57 progress=1.000000  cache=0.2MiB(536tx)
AddToWallet 128e13f592d42b014181f92b8ba3ac395bd50238071a78c59dc6eb733b179f89  
zcashd: wallet/wallet.cpp:605: void CWallet::IncrementNoteWitnesses(const CBlockIndex*, const CBlock*, ZCIncrementalMerkleTree): Assertion `nWitnessCacheSize >= nd->witnesses.size()' failed.
Aborted
```

This is after my VM crashed and restarted; don't know whether that's relevant.

### PR DESCRIPTION
**Pull request by str4d**
Closes #1378

### CODE REVIEW

**Comment by daira** [comment]
@str4d You said something about checking the height in the case that had the bug; I don't think we implemented that.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -708,9 +708,11 @@ void CWallet::IncrementNoteWitnesses(const CBlockIndex* pindex,
...
                                 // doesn't affect existing cached notes because of the
                                 // CNoteData::witnessHeight checks. See #1378 for details.
-                                LogPrintf("Inconsistent witness cache state found for %s\n- Cache size: %d\n- Top: %s\n- New: %s\n",
+                                LogPrintf("Inconsistent witness cache state found for %s\n- Cache size: %d\n- Top (height %d): %s\n- New (height %d): %s\n",
                                           jsoutpt.ToString(), nd->witnesses.size(),
+                                          nd->witnessHeight,
                                           nd->witnesses.front().root().GetHex(),
+                                          pindex->nHeight,
                                           tree.witness().root().GetHex());
                                 nd->witnesses.clear();
```
Comment (daira) [inline src/wallet/wallet.cpp]: Does ``nd->witnessHeight`` need to be changed when we clear the cache?

**Comment by str4d** [comment]
Hmm, reposting here because GitHub thinks the new commit is outdated...

@daira
> Does nd->witnessHeight need to be changed when we clear the cache?

No, because we set it immediately below.

**Comment by daira** [comment]
ut(ACK+cov). Needs another ACK. (I paired with @str4d on implementing this.)

**Comment by bitcartel** [comment]
@str4d @daira This is in regtest mode, when trying to re-launch node.  I launced a new regtest node, generated 110 blocks, created a z_sendmany (taddr to zaddr), generated 1 block, then quit.  On relaunch:
```
zcashd: wallet/wallet.cpp:702: void CWallet::IncrementNoteWitnesses(const CBlockIndex*, const CBlock*, ZCIncrementalMerkleTree&): Assertion `nd->witnesses.size() == 0' failed.
```
commit: b492ef323230f577636a16d2bb156796ad63f794
Auto merge of #1568 - str4d:1548-remove-block-index-tag, r=daira
Remove the RC 1 block index error message

**Comment by str4d** [comment]
@bitcartel that should be fixed by this PR.

**Comment by bitcartel** [comment]
@str4d Using this PR, lauching zcashd, the assertion was no longer triggered and my unspent note was found and could be spent.

**Comment by bitcartel** [comment]
ACK (did not check coverage)

**Comment by daira** [comment]
@zkbot r+

**Comment by str4d** [comment]
@zkbot retry

**Comment by str4d** [comment]
Manually cancelled previous merge so servers could be restarted with security patches.

@zkbot retry

**Pull request merged**